File: AccuPay/Core/Utilities/Extensions/ObjectExtensions.vb

Imports System.Runtime.CompilerServices
Imports Newtonsoft.Json

Namespace Global.AccuPay.Extensions

    Module ObjectExtensions

        <Extension()>
        Public Function CloneJson(Of T)(ByVal source As T) As T
            Return CloneObject(source)
        End Function

        <Extension()>
        Public Function CloneListJson(Of T)(ByVal source As List(Of T)) As List(Of T)

            Dim newList As New List(Of T)

            For Each item In source
                newList.Add(item.CloneJson())
            Next

            Return newList

        End Function

#Region "Private Functions"

        Private Function CloneObject(Of T)(source As T) As T
            If Object.ReferenceEquals(source, Nothing) Then
                Return Nothing
            End If

            Dim deserializeSettings = New JsonSerializerSettings With {
                .ObjectCreationHandling = ObjectCreationHandling.Replace
            }
            Return JsonConvert.DeserializeObject(Of T)(JsonConvert.SerializeObject(source), deserializeSettings)
        End Function

    End Module
#End Region

End Namespace

File: AccuPay.Utilities/Extensions/ObjectExtensions.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace AccuPay.Utilities.Extensions
{
    public static class ObjectExtensions
    {
        public static T CloneJson<T>(this T source)
        {
            return CloneObject(source);
        }

        public static List<T> CloneListJson<T>(this List<T> source)
        {
            var newList = new List<T>();

            if (source == null) return newList;

            foreach (var item in source)
            {
                newList.Add(item.CloneJson());
            }

            return newList;
        }

        public static bool NullableEquals<T>(this T source, T comparedTo)
        {
            // if both are null, they are equal
            if (source == null && comparedTo == null)
                return true;

            // if one is null and the other is not, they are not equal
            if ((source == null && comparedTo != null) || (source == null && comparedTo != null))
                return false;

            // if both have value, compare them
            return source.Equals(comparedTo);
        }

        private static T CloneObject<T>(T source)
        {
            if (Object.ReferenceEquals(source, null)) return default(T);

            var deserializeSettings = new JsonSerializerSettings()
            {
                ObjectCreationHandling = ObjectCreationHandling.Replace
            };

            return JsonConvert.DeserializeObject<T>(JsonConvert.SerializeObject(source),
                                                    deserializeSettings);
        }
    }
}